base/opaque_closure.jl

```diff
@@ -24,3 +24,46 @@ end
 macro opaque(ty, ex)
     esc(Expr(:opaque_closure, ty, ex))
 end
+
+# OpaqueClosure construction from pre-inferred CodeInfo/IRCode
+using Core.Compiler: IRCode
+using Core: CodeInfo
+
+function compute_ir_rettype(ir::IRCode)
+    rt = Union{}
+    for i = 1:length(ir.stmts)
+        stmt = ir.stmts[i][:inst]
+        if isa(stmt, Core.Compiler.ReturnNode) && isdefined(stmt, :val)
+            rt = Core.Compiler.tmerge(Core.Compiler.argextype(stmt.val, ir), rt)
+        end
+    end
+    return Core.Compiler.widenconst(rt)
+end
+
+function Core.OpaqueClosure(ir::IRCode, env...;
+        nargs::Int = length(ir.argtypes)-1,
```

Comment:
Shouldn't `nargs` be a required argument? I think `length(ir.argtypes)-1` doesn't necessarily correspond to the number of actual argument as it can include entries for arbitrary slots.

Reply:
Conceptually only the ir.argtypes that correspond to the arguments have any meaning. We sometimes leave the slottypes in there for convenience, but we should probably stop doing that.